File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanTest.java
package rocks.frieler.android.beans.scopes.singleton;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java8.util.function.Supplier;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBean.lazy;
import static rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBeanHandler.SINGLETON_SCOPE;

@RunWith(MockitoJUnitRunner.class)
public class SingletonScopedFactoryBeanTest {

    @SuppressWarnings("unchecked")
    private Supplier<SingletonScopedFactoryBeanTest> producer = mock(Supplier.class);
    private SingletonScopedFactoryBean<SingletonScopedFactoryBeanTest> factoryBean = lazy(SingletonScopedFactoryBeanTest.class, producer);

    @Test
    public void testGetScopeReturnsSingletonScope() {
        assertThat(factoryBean.getScope(), is(SINGLETON_SCOPE));
    }

    @Test
    public void testGetTypeReturnsConfiguredType() {
        assertThat(factoryBean.getBeanType(), is(equalTo(SingletonScopedFactoryBeanTest.class)));
    }

    @Test
    public void testProduceBeanCallsConfiguredProducer() {
        when(producer.get()).thenReturn(this);

        assertThat(factoryBean.produceBean(), is(this));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/singleton/SingletonScopedFactoryBeanTest.kt
package rocks.frieler.android.beans.scopes.singleton

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isInstanceOf
import assertk.assertions.isSameAs
import assertk.assertions.isSameInstanceAs
import org.junit.jupiter.api.Test
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import rocks.frieler.android.beans.BeanDefinition
import rocks.frieler.android.beans.BeansProvider
import rocks.frieler.android.beans.DeclarativeBeanConfiguration
import rocks.frieler.android.beans.scopes.ScopedBeanDefinition
import rocks.frieler.android.beans.scopes.singleton.SingletonScopedFactoryBean.Companion.lazyInstantiated

class SingletonScopedFactoryBeanTest {
    private val producer: BeansProvider.() -> SingletonScopedFactoryBeanTest = mock()
    private val factoryBean = SingletonScopedFactoryBean(SingletonScopedFactoryBeanTest::class, producer)

    @Test
    fun `scope() is singleton scope`() {
        assertThat(factoryBean.scope).isEqualTo(SingletonScopedFactoryBeanHandler.SINGLETON_SCOPE)
    }

    @Test
    fun `beanType is the configured type`() {
        assertThat(factoryBean.beanType).isEqualTo(SingletonScopedFactoryBeanTest::class)
    }

    @Test
    fun `produceBean() calls the configured producer`() {
        val dependencies = mock<BeansProvider>()
        whenever(producer(dependencies)).thenReturn(this)

        val producedBean = factoryBean.produceBean(dependencies)

        verify(producer).invoke(dependencies)
        assertThat(producedBean).isSameInstanceAs(this)
    }

    @Test
    fun `lazyInstantiated() without dependencies provides a BeanDefinition for a SingletonScopedFactoryBean`() {
        val dependencies = mock<BeansProvider>()
        val producerWithoutDependencies: () -> SingletonScopedFactoryBeanTest = mock()
        whenever(producerWithoutDependencies()).thenReturn(this)

        val lazyInstantiatedBeanDefinition = lazyInstantiated(SingletonScopedFactoryBeanTest::class.java, producerWithoutDependencies)

        assertThat(lazyInstantiatedBeanDefinition).isInstanceOf(ScopedBeanDefinition::class)
        assertThat(lazyInstantiatedBeanDefinition.getType()).isEqualTo(SingletonScopedFactoryBean::class)
        assertSingletonScopedFactoryBeanProducingThis(lazyInstantiatedBeanDefinition.produceBean(dependencies), dependencies)
    }

    @Test
    fun `lazyInstantiated() provides a BeanDefinition for a SingletonScopedFactoryBean`() {
        val dependencies = mock<BeansProvider>()
        whenever(producer(dependencies)).thenReturn(this)

        val lazyInstantiatedBeanDefinition = lazyInstantiated(SingletonScopedFactoryBeanTest::class.java, producer)

        assertThat(lazyInstantiatedBeanDefinition).isInstanceOf(ScopedBeanDefinition::class)
        assertThat(lazyInstantiatedBeanDefinition.getType()).isEqualTo(SingletonScopedFactoryBean::class)
        assertSingletonScopedFactoryBeanProducingThis(lazyInstantiatedBeanDefinition.produceBean(dependencies), dependencies)
    }

    @Test
    fun `DeclarativeBeanConfiguration_lazyInstantiatedBean() declares a SingletonScopedFactoryBean`() {
        val dependencies = mock<BeansProvider>()
        val beanConfiguration : DeclarativeBeanConfiguration = mock()
        val beanName = "aLazyInstantiatedBean"
        whenever(producer(dependencies)).thenReturn(this)

        beanConfiguration.lazyInstantiatedBean(beanName) { producer() }

        val beanDefinitionCaptor = argumentCaptor<BeanDefinition<SingletonScopedFactoryBean<SingletonScopedFactoryBeanTest>>>()
        verify(beanConfiguration).addBeanDefinition(beanDefinitionCaptor.capture())
        val beanDefinition = beanDefinitionCaptor.firstValue
        assertThat(beanDefinition).isInstanceOf(ScopedBeanDefinition::class)
        assertThat(beanDefinition.getName()).isEqualTo(beanName)
        assertThat(beanDefinition.getType()).isEqualTo(SingletonScopedFactoryBean::class)
        assertSingletonScopedFactoryBeanProducingThis(beanDefinition.produceBean(dependencies), dependencies)
    }

    private fun assertSingletonScopedFactoryBeanProducingThis(factoryBean: SingletonScopedFactoryBean<*>, dependencies: BeansProvider) {
        assertThat(factoryBean).isInstanceOf(SingletonScopedFactoryBean::class)
        assertThat(factoryBean.beanType).isEqualTo(SingletonScopedFactoryBeanTest::class)
        assertThat(factoryBean.produceBean(dependencies)).isSameInstanceAs(this)
    }
}
